content/en/blog/_posts/2022-04-19-dockershim-historical-context.md

```diff
@@ -0,0 +1,23 @@
+---
+layout: blog
+title: "Dockershim: The Historical Context"
+date: 2022-04-19
+slug: dockershim-historical-context
+---
+
+**Author:** Kat Cosgrove
+
+
+Dockershim has been removed as of Kubernetes v1.24, and this is overall a positive move for the project. However, context is important for fully understanding something, be it socially or in software development, and this deserves a more in depth review. Alongside the dockershim removal in Kubernetes v1.24, we’ve seen some confusion and even panic in the community, largely a result of a lack of context around this removal. The decision to deprecate and eventually remove it from Kubernetes was not made quickly or lightly, but it’s been in the works for so long that many of today’s users are newer than that decision, and certainly newer than the choices that led to the dockershim being necessary in the first place.
+
+So what is the dockershim, and why is it going away?
+
+In the early days of Kubernetes, we only supported one container runtime. That runtime was Docker Engine. Back then, there weren’t really a lot of other options out there and Docker was the dominant tool for working with containers, so this was not a controversial choice. Eventually, we started adding more container runtimes, like rkt and hypernetes, and it became clear that Kubernetes needed a way to allow cluster operators the flexibility to use whatever runtime they choose.
+
+The [Container Runtime Interface](/blog/2016/12/container-runtime-interface-cri-in-kubernetes/) (CRI) was released to allow that flexibility. The introduction of CRI was great for the project and users alike, but it did introduce a problem: Docker Engine’s use as a container runtime predates CRI, and Docker Engine is not CRI-compatible. To solve this issue, a small software shim (dockershim) was introduced specifically to fill in the gaps between Docker Engine and CRI, allowing cluster operators to continue using Docker Engine as their container runtime largely uninterrupted.
+
+This little software shim was never intended to be a permanent solution, though. Over the course of years, its existence has introduced a lot of unnecessary complexity to the kubelet itself. Some integrations are inconsistently implemented for Docker because of this shim, resulting in an increased burden on maintainers, and maintaining vendor-specific code is not in line with our open source philosophy. To reduce this maintenance burden and move towards a more collaborative community in support of open standards, [KEP-2221 was introduced](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2221-remove-dockershim), proposing the removal of the dockershim. With the release of Kubernetes v1.21, the deprecation was official.
```

Comment:
Dockershim was deprecated in v1.20
See https://kubernetes.io/blog/2022/02/17/dockershim-faq/

```suggestion
This little software shim was never intended to be a permanent solution, though. Over the course of years, its existence has introduced a lot of unnecessary complexity to the kubelet itself. Some integrations are inconsistently implemented for Docker because of this shim, resulting in an increased burden on maintainers, and maintaining vendor-specific code is not in line with our open source philosophy. To reduce this maintenance burden and move towards a more collaborative community in support of open standards, [KEP-2221 was introduced](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2221-remove-dockershim), proposing the removal of the dockershim. With the release of Kubernetes v1.20, the deprecation was official.
```